### ISSUE TITLE
How to initialize the weight of each letter separately for ASR (QuartzNet and CitriNet)?

### ISSUE THREAD

**Issue by psydok** [NVIDIA/NeMo#4288]
I want to do Transfer Learning by explicitly initializing the en -> ru alphabet.
```python
dict_vocab= {
     "a": "а",
     "b": "б",
     "c": "с",
     "d": "д",
     "f": "ф",
...
}
```

Do not tell me how this can be done? I didn't find it in the tutorials.
Thanks in advance!

**Comment by titu1994** [comment]
If it's a character based model like QuartzNet, you can just pass the list of values in your dict to change_vocabulart(). 

Subword models like Citrinet do not support such mapping, you can just use Tokenizers with the Russian alphabet and the model should learn the new token mapping.

**Comment by psydok** [comment]
@titu1994  I want to do a weight transfer of each character. And so I get to reset the decoder, and I will have to train from scratch, if I understand correctly.

**Comment by titu1994** [comment]
Yes, as long as the characters map 1:1, you can load the weights. If there is not 1;1 map, you have to load the weights matrix and manually index the weight and stitch them into the new model decoder.

**Comment by psydok** [comment]
And how can I check that the desired weights have loaded?
The only tip when I do:
```python
model.encoder = old_model.encoder
model.decoder = old_model.decoder
```
The quality of recognition becomes at the level of the old_model of the model, but I can’t confirm for myself that I transferred the weights and how.

**Comment by titu1994** [comment]
There's no code to do it as you exactly need it. Everything is pytorch, so you can code if yourself if you need it. You can access the decoder module using model.decoder, and use new_model.decoder.load_state_dict(old_model.decoder.state_dict()) if the shapes match. 

If they don't you will need to modify the state dict itself to chop off the tensors that don't match the new decoder shape. 

Btw, we don't recommend doing any of this. We have Russian models that are finetuned from English alphabet (both char and subword based) and find that there is no need for any such language specific weight stitching. It does not speed up convergence, if does not lead to better or worse scores, and it's nearly redundant for subword based models where you can chose to use same size of vocabulary (128 or 1024 depending on model) and then load up the entire English model decoder weights to get fast convergence and better wer. 

What I mean to say is such language specific 1:1 napping of English and Russian alphabet mapping is not at all significantly improving wer or speed of convergence.

**Comment by titu1994** [comment]
No. There is no need to retrain in 8k. We find 16 khz model works better after upsampling 8khz audio, than having a true 8khz model.

**Comment by psydok** [comment]
Good. Thanks, I saw!

I would also like to try to train citrinet, but when I try to initialize the voice model decoder weights (new_model.decoder.load_state_dict(old_model.decoder.state_dict())), it tells me that the size of the dictionary does not match. Indeed, the spe(char) tokenizer has added another character to the dictionary (`<unk>`).

I understand correctly, that each weight from `golos_model.decoder.state_dict()["decoder_layers.0.weight"]` corresponds to labels in order? That is, `golos_model.decoder.state_dict()["decoder_layers.0.weight"][1]` == weight of letter `a`?
If the labels of the quartz voice model looks like this:
```
  labels: &labels [ " ", "а", "б", "в", "г", "д", "е", "ж", "з", "и", "й", "к", "л", "м", "н", "о", "п",
                    "р", "с", "т", "у", "ф", "х", "ц", "ч", "ш", "щ", "ъ", "ы", "ь", "э", "ю", "я" ]
```

**Comment by psydok** [comment]
I figured it out, I didn’t notice right away that it was also necessary to add a value to bias. 
I added a zero to the beginning.

**Comment by psydok** [comment]
Could you also tell me if it will be possible to achieve good results on citrinet in this way for training a recognition model on datasets of different domains + golos?

**Comment by titu1994** [comment]
You don't really need to stitch such weights together. There's no need with Tokenizer based models. We found that we simply fine tuned a Citrinet with 1024 tokens in English, replaced the decoder with 1024 token of Russian subwords, loaded English decoder weights and finetuned on Russian corpus. This gave us excellent scores as I've mentioned before. 

We didn't need to do any of this stitching of weights. We have not tried such a method so we can't comment on if it will work or not.

**Comment by psydok** [comment]
Thanks a lot! Tell me, please, on what volume of the corpus did you retrain the English-language model into Russian? How much time did it take you?
I tried to put it on the GOLOS case with batch size 32 from scratch. It is trained very slowly, there is no quality yet.
Although I haven't tried your method yet, but there were doubts whether it was worth trying to train such a model at all. Initially, the Citrinet model was attractive because, judging by the fact that it is non-autoregressive, it should be faster than QuartzNet.
But is it possible to tweak the adaptation to the domain to it (although the same question is for QuartzNet)?
And will the mechanics of streaming recognition be the same as in the QuartzNet tutorial? 
As far as I understand, then even when working with a stream, it is necessary to accumulate at least 4-8 seconds for the model to recognize the text with minimal loss of quality. 
Perhaps Citrinet can do a better job of doing this with less need to accumulate data?

**Comment by titu1994** [comment]
If you need streaming models, or at least strong streaming performance with offline models, I think Conformer will be much better choice than either of the above models. As to question regarding wer on Russian corpus, I'll ask @vsl9 to commentv
